File: python/leetcode152.py
class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #https: // leetcode.com / problems / maximum - product - subarray / discuss / 210384 / Python - DP - Time - O(N) - Space - O(1)

        promax = promin = res = nums[0]
        for num in nums[1:]:
            if num < 0:
                promax, promin = promin, promax

            promax = max(num, num*promax)
            promin = min(num, num*promin)
            res = max(res, promax)
        return res

File: leetcodec++/leetcode152.cpp
//
// Created by 黄敏辉 on 2019-08-21.
//

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int front = 1;
        int back = 1;
        int res = INT_MIN;
        int l = nums.size();
        for (int i = 0; i < l; i++){
            front *= nums[i];
            back *= nums[l - 1 - i];
            res = max(res, max(front, back));
            front = front == 0 ? 1 : front;
            back = back == 0 ? 1 : back;
        }
        return res;
    }
};